### PR TITLE
Add is_inference to native functions

### PR DESCRIPTION
**Pull request by soulitzer**
Adds `is_inference` as a native function w/ manual cpp bindings.
Also changes instances of `is_inference_tensor` to `is_inference` to be consistent with other properties such as `is_complex`.

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -1313,7 +1313,7 @@ std::vector<Tensor> unsafe_split(const Tensor& self, int64_t split_size, int64_t
   auto result = at::native::split(self, split_size, dim);
   for (auto& t : result) {
     // TODO(Ailing): do we need to set version_counter here?
-    if (!t.unsafeGetTensorImpl()->is_inference_tensor()) {
+    if (!t.unsafeGetTensorImpl()->is_inference()) {
```
Comment (ailzhang) [inline aten/src/ATen/native/TensorShape.cpp:1316]: After the change there will be a `is_inference` method on `Tensor` as well, so I think it's safe to remove most `unsafeGetTensorImpl()` calls :D

Comment (soulitzer) [reply aten/src/ATen/native/TensorShape.cpp:1316]: Good point!

File: torch/_tensor_docs.py [hunk 2]
```diff
@@ -1970,6 +1970,13 @@ def add_docstr_all(method, docstr):
 Returns True if the data type of :attr:`self` is a complex data type.
 """)
 
+add_docstr_all('is_inference',
+               r"""
+is_inference() -> bool
+
+Returns True if :attr:`self` is an infernece tensor.
```
Comment (ailzhang) [inline torch/_tensor_docs.py:1977]: typo: infernece ;)

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -4185,6 +4185,15 @@ def merge_dicts(*dicts):
 Returns True if inference mode is currently enabled.
 """.format(**common_args))
 
+add_docstr(torch.is_inference, r"""
+is_inference(input) -> (bool)
+
+Returns True if :attr:`input` is an inference tensor
```
Comment (ailzhang) [inline torch/_torch_docs.py:4191]: I'd suggest adding a link to the note where we explain what's an inference tensor (if we already have one). If not we probably need to explain more here :D

**Review by ailzhang — APPROVED** [review]
Looks great! Only have a few minor suggestions.

File: torch/_torch_docs.py (hunk 3 above)
Comment (soulitzer) [reply torch/_torch_docs.py:4191]: How about something like:
```
Returns True if :attr:`input` is an inference tensor.

A non-view tensor is an inference tensor if and only if it was
allocated during inference mode. A view tensor is an inference
tensor if and only if the tensor it is a view of is an inference tensor. <-- From RFC

For details on inference mode please see
`Inference Mode <https://pytorch.org/cppdocs/notes/inference_mode.html>`_.
```

Comment (ailzhang) [reply torch/_torch_docs.py:4191]: LGTM!

**Pull request closed without merging**